### ISSUE TITLE
naming-convention: treat variables initialized with an arrow function as functions

### ISSUE THREAD

**Issue by reduckted** [ajafff/tslint-consistent-codestyle#41]
Here's my `naming-convention` configuration:

```json
"rules": {
    "naming-convention": [ true,
      { "type": "default", "format": "camelCase" },
      { "type": "variable", "modifiers": "export", "format": "UPPER_CASE" },
      { "type": "function", "modifiers": "export", "format": "camelCase" }
    ]
}
```

As you can see from this, I want exported variables to be upper case, but exported functions to be camel case.

Here's some typical code that I'd write (it's ngrx-style code):

```typescript
export const ACTION = "submit";
export const getValue = (state: State) => state.value;
```

This fails linting because getValue is expected to be upper case, but I'd prefer it to be camel case. I _could_ write it like this:

```typescript
export function getValue(state: State): boolean { return state.value; }
```

But it's a bit more verbose and (because of other linting rules I use) requires the function to have a type annotation.

Would it be possible to treat variables that are initialized with an arrow function to be treated as functions instead of variables?

I've taken a look at the rule, and it's simple enough to check if a variable is initialized to an arrow function:

```typescript
if (declaration.initializer && declaration.initializer.kind === ts.SyntaxKind.ArrowFunction) {
    this._checkName(declaration.name, TypeSelector.function, currentModifiers);
} else {
    this._checkName(declaration.name, TypeSelector.variable, currentModifiers);
}
```
But I'm wondering if maybe it should be an option to make it behave this way so that it doesn't break everyone else's existing configurations.

**Comment by ajafff** [comment]
I guess it makes sense to add a new type `functionVariable` that matches variables initialized with an arrow function or function expression. It will not use type information to infer the type of the initializer.
